core/src/test/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortIT.java

```diff
@@ -0,0 +1,348 @@
+                        .subAggregation(bucketSort("bucketSort", Arrays.asList(
+                                new FieldSortBuilder("_count").order(SortOrder.ASC),
+                                new FieldSortBuilder("avg_value").order(SortOrder.DESC)))))
+                .execute().actionGet();
+
+        assertSearchResponse(response);
+
+        Terms terms = response.getAggregations().get("foos");
+        assertThat(terms, notNullValue());
+        List<? extends Terms.Bucket> termsBuckets = terms.getBuckets();
+        long previousCount = termsBuckets.get(0).getDocCount();
+        double previousAvgValue = ((Avg) termsBuckets.get(0).getAggregations().get("avg_value")).getValue();
+        for (Terms.Bucket termBucket : termsBuckets) {
+            Avg avg = termBucket.getAggregations().get("avg_value");
+            assertThat(avg, notNullValue());
+            assertThat(previousCount, lessThanOrEqualTo(termBucket.getDocCount()));
+            if (previousCount == termBucket.getDocCount()) {
+                assertThat(previousAvgValue, greaterThanOrEqualTo(avg.getValue()));
+            }
+            previousCount = termBucket.getDocCount();
+            previousAvgValue = avg.getValue();
+        }
+    }
+
+    public void testSortDateHistogramDescending() {
+        SearchResponse response = client().prepareSearch(INDEX)
+                .addAggregation(dateHistogram("time_buckets").field(TIME_FIELD).interval(TimeValue.timeValueHours(1).millis()))
+                .execute().actionGet();
+
+        assertSearchResponse(response);
+
+        Histogram histo = response.getAggregations().get("time_buckets");
+        assertThat(histo, notNullValue());
+        assertThat(histo.getName(), equalTo("time_buckets"));
+        List<? extends Histogram.Bucket> ascendingTimeBuckets = histo.getBuckets();
+
+        response = client().prepareSearch(INDEX)
+                .addAggregation(dateHistogram("time_buckets").field(TIME_FIELD).interval(TimeValue.timeValueHours(1).millis())
+                        .subAggregation(bucketSort("bucketSort", Arrays.asList(
+                                new FieldSortBuilder("_key").order(SortOrder.DESC)))))
+                .execute().actionGet();
+
+        assertSearchResponse(response);
+
+        histo = response.getAggregations().get("time_buckets");
+        assertThat(histo, notNullValue());
+        assertThat(histo.getName(), equalTo("time_buckets"));
+        List<? extends Histogram.Bucket> descendingTimeBuckets = histo.getBuckets();
+
+        assertThat(ascendingTimeBuckets.size(), equalTo(descendingTimeBuckets.size()));
+        int bucketCount = ascendingTimeBuckets.size();
+        for (int i = 0; i < bucketCount; ++i) {
+            assertThat(ascendingTimeBuckets.get(i), equalTo(descendingTimeBuckets.get(bucketCount - i - 1)));
+        }
+    }
+
+    public void testSortHistogram_GivenGapsAndGapPolicyIsSkip() {
```

Comment:
Re: size + skipping issue from above, might be good to have a test for that scenario where an explicit size is set (which is less than the available buckets but spans across a skip). 

Skips make everything more complicated :(